sampleSTAT.1
.\"Created with GNOME Manpages Editor Wizard
.\"http://sourceforge.net/projects/gmanedit2
.TH sampleSTAT 1 "November 7, 2020" "" "sampleSTAT"

.SH NAME
sampleSTAT \- Statistical analysis of univariate data

.SH SYNOPSIS
.B sampleSTAT
[\fB\-hv\fR] [\fb\-l\fR \fICONFIDENCE_LEVEL\fR]  \fI\,[<DATASOURCE] [>RESULT]\/\fR
.br

.SH DESCRIPTION
\fBsampleSTAT\fP provides elementary tests for the evaluation of univariate measurement data that are typically recorded by scientists and engineers. These data have to be normally distributed.

It provides calculations of the range of values and the mean with respect to a given statistical confidence level. 

The following tests will be performed:

- Arithmetic Mean
.br
- Sample Standard Deviation
.br
- Range of Dispersion of Values (stray area of the values)
.br
- Range of Dispersion of Mean (trust area of the mean)
.br
- Minimum Value
.br
- Maximum Value

Find a description of the tests below.

.SH OPTIONS
.TP
\fB\-l\fR \fICONFIDENCE_LEVEL_(P)\fR, \fB\-\-level\fR=\fICONFIDENCE_LEVEL_(P)\fR
Statistical confidence level P (95%, 99% or 99.9%):
.br
P=95    =>   confidence level: 95%
.br
P=99    =>   confidence level: 99%
.br
P=99.9  =>   confidence level: 99.9%
.TP
\fB\-h\fR, \fB\-\-help\fR
Print help screen
.TP
\fB\-v\fR, \fB\-\-version\fR
Print version information
.TP
\fI\<DATASOURCE\fR
Redirection of the datasource, e.g. a datafile, to \fIstdin\fR for reading. This option is optional. 

You can omit this option and type the data to \fIstdin\fR instead. When you finished you press CTRL-D.
.TP
\fI\>RESULT\fR
Redirection of the result to e.g. a file. This option is optional. Otherwise \fBsampleSTAT\fR will print the result to the screen.

.SH EXAMPLES
sampleSTAT -l 95 <mydata.dat
.br
sampleSTAT --level=99 <mydata.dat >results.txt
.br
sampleSTAT -l 99.9 <mydata.dat
.br
sampleSTAT -v
.br
sampleSTAT -vh

.SH INPUT DATA
Data can be committed via file (redirection to \fIstdin\fR), pipe or keyboard. The format of the data has to be 1-dimensional \fIcomma-separated-value\fR. Committed  as a row vector and the decimal separator has to be a point, e.g:

22.43
.br
22.45
.br
22.50

.SH OUTPUT
Print the results of the perfomed calculations in a table to \fIstdout\fR, e.g.:

 sampleSTAT - Statistics for Sampling Distributions
.br
 ==================================================
.br
 Number of Values            :    6
.br
 Arithmetic Mean             :    10.000000000000000     
.br
 Confidence Level            :    95%
.br
 Range of Dispersion (values):    3.6345288507107026E-003
.br
 Range of Dispersion (mean)  :    1.4837901899442333E-003
.br
 Standard Deviation          :    1.4142135605878221E-003
.br
 Minimum                     :    9.9979999999999993     
.br
 Maximum                     :    10.002000000000001  

.SS INTERPRETATION

Write as a result:
.br
x = 10.000; T = +/-0.0036 for P = 95%; s = +/-0.0014

Result in words: 
.br
In the given sample distribution 95% of all values are expected between 9.9964 and 10.0036 (range of dispersion of the values) but 68% between 9.9986 and 10.0014 (standard deviation). The mean "x" itself can spread between 9.9985 and 10.0015 (range of dispersion of the mean) with a probability of 95%.

.SH DESCRIPTION OF THE STATISTICAL RESULTS
If you are collecting data on a process, it is important to determine not only the location of the mean, but also to look at the variation within the data. If you are, for example, interpreting the results of a chemical analysis, you may put much more emphasis on the obtained average value if you know that the individual samples vary only very little in comparison to the mean.

In general, the spread of a distribution, both in absolute and in relative terms, is a good measure of the variability (and hence reliability) of the data. There are several ways to specify the variation in the data. \fBsampleSTAT\fR provides you more information of your data as just mean, median and standard deviation can do.
.SS
Arithmetic Mean (x)
The arithmetic mean is the best and significant approach to specify an result when it will used numerically. For its alone it does not say anything about the quality of the result. An artithmetic mean of 10 can be generated from both examples below:

Sample 1 	Sample 2
.br
6 	     9.999
.br
8 	     9.998
.br
14 	    10.002
.br
12 	    10.000
.br
5        10.001
.br
15       10.000
.br
.SS
Confidence Level
The confidence level is the percentage of times you expect to reproduce an estimate between the upper and lower bounds of the confidence interval.
.SS
Number of Values (N)
The amount of single values (N) is a important degree to evaluate the security of the result. N specify the measuring expenditure but N itself has no significance to the the result. In both examples above N was the same.
.SS
Sample Standard Deviation (s)
The standard deviation is the most informative quantity to specify the statistic quality of a measuring procedure. With a probability of 68% the values will stray with +/- s around the mean. But s itself does not say anything about the quality of the result but in combination with n you can calculate the range of dispersion.
.SS
Range of Dispersion of Values (T) - stray area of the values
The range of dispersion of the values is calculated as T = s * t. t is the student factor, dependent on the statistic security/confidence level P% (95%, 99%, 99.9%) and the degree of freedom f = n - 1. f and t are specified in the t-tables. T indicates that P% of all single values - with x as its mean - are expected around x + T and x - T. T is the parameter which specify the quality of the raw values.
.SS
Range of Dispersion of Mean - trust area of the mean
This parameter tells how secure the mean is. It indicates the stray area of the mean and not of the raw values as T does. The mean will stray with a probability of P% around the mean.

.SH AUTHOR
sampleSTAT was written by Hani Ibrahim.

.SH
BIBLIOGRAPHY
R. Kaiser, G. Gottschalk; "Elementare Tests zur Beurteilung von Meßdaten", BI Hochschultaschenbücher, Bd. 774, Mannheim 1972.

.SH LICENSE
GNU Public License Version 3 - \<http://www.gnu.org/licenses/\>



sampleSTAT.f90
! sampleSTAT - Statistics for Sampling Distributions
! =============================================================================
!
! Purpose:
!    Performs statistical tests for sampling distributions:
!      - Arithmetic Mean
!      - Standard Deviation
!      - Range of Dispersion of Values
!      - Range of Dispersion of Mean
!      - Minimum Value
!      - Maximum Value
!
!
! History:
!    Version   Programmer         Date       Description
!    -------   ----------         ---------- -----------
!    1.0       Ibrahim, Hani      2006-05-29 Original code
!    1.1       Ibrahim, Hani      2016-01-19 Changed -s/--sens to -l/--level
!                                            and some little cosmetic corrections
!    1.2       Ibrahim, Hani      2020-11-07 Switched to linked list for reading data
!
! Processing:
!   Read 1-dimensional numerical data from stdin (pipe/redirection) and calculate
!   elementary tests for the evaluation of univariate measurement data that are
!   typically recorded by scientists and engineers. These data have to be normally
!   distributed for sampleSTAT's routines:
!

MODULE PrgMod
    USE SysConst
    USE SampleStatistics

    IMPLICIT NONE
	
	INTEGER :: stderr ! stderr unit

CONTAINS

!----------------------------------------------------------------------
   SUBROUTINE Foo()
      IMPLICIT NONE
      WRITE(StdErr,*) 'Option not implemented yet'
   END SUBROUTINE Foo
!----------------------------------------------------------------------
   SUBROUTINE Version()
   ! Write version information to stdout
      IMPLICIT NONE
      WRITE(*,'(/,5(A/))') &
      'sampleSTAT - Version 1.2 - 2020-11-07',&
      'For information, please contact: Hani Andreas Ibrahim (hani.ibrahim@gmx.de)',&
      'sampleSTAT comes with NO WARRANTY, to the extent permitted by law. You may ',&
      'redistribute copies of sampleSTAT under the terms of the GNU General Public ',&
      'License, refer <www.gnu.org/licenses/gpl.html> for details.'
   END SUBROUTINE Version
!----------------------------------------------------------------------
   SUBROUTINE Help()
   ! Write help page to stdout
      WRITE(*,'(/,3(A/),/,7(A/),/,(4(A/)),/,(5(A/)),/)')&
      'sampleSTAT performs tests for statistical samples:',&
      '   Aritmetic Mean, Range of Dispersion of values and mean based on t-factor,', &
      '   Standard Deviation, Minimum, Maximum.', &
      'Usage: sampleSTAT [-hv] [-l P] [<inputfile] [>outputfile]',&
      '  -h    --help /?   Print this help screen',&
      '  -v    --version   Print version information',&
      '  -l P  --level=P   Set confidence level:',&
      '                    P=95   => conf. level: 95%',&
      '                    P=99   => conf. level: 99%',&
      '                    P=99.9 => conf. level: 99.9%',&
      'Examples:',&
      '  sampleSTAT -l 95 <mydata.dat',&
      '  sampleSTAT --level=99 <mydata.dat >results.txt',&
      '  sampleSTAT -l 99.9 <mydata.dat',&
      'Input data:',&
      '  Data has to be committed in a one column form, like:',&
      '     22.43',&
      '     22.45',&
      '     22.50'
   END SUBROUTINE Help

END MODULE PrgMod

!======================================================================

PROGRAM sampleSTAT

   USE SysConst
   USE SampleStatistics
   USE PrgMod     ! ReadInVec
   USE Sng        ! [libsng] Command-line parsing
   USE ReadData   ! Read data from stdin

   IMPLICIT NONE

   INTEGER                                  :: I                 ! Loop index
   INTEGER                                  :: ReadErr           ! error variables
   REAL(KIND=DP), DIMENSION(:), POINTER     :: Values_ptr        ! Data values
   INTEGER                                  :: N                 ! Numbers of values
   REAL(KIND=SP)                            :: Level = -1.       ! confidence level ...
                                                                 ! ... for "OutlierOut" routine
   CHARACTER(LEN=6)                         :: LevelMsg          ! confidence level string
   REAL(KIND=DP)                            :: StrayAreaResult   ! Stray Area of single values depending on conf. level
  ! Command-line variables
   character(16)                            ::arg_val            ! [sng] Command line argument value
   character(16)                            ::opt_sng            ! [sng] Option string
   character(2)                             ::dsh_key1, dsh_key2 ! [sng] Command line dash and switch
   integer                                  ::arg_idx            ! [idx] Counting index
   integer                                  ::arg_nbr            ! [nbr] Number of command line arguments
   integer                                  ::opt_lng            ! [nbr] Length of option

   ! Command-line option switches
   LOGICAL                                  :: LogLevel =.FALSE. ! Value "Level" commited: true/false
   LOGICAL                                  :: LogHlp   =.FALSE. ! Switch "help"
   LOGICAL                                  :: LogVer   =.FALSE. ! Switch "version"


   ! Error messages
   CHARACTER(Len=*), PARAMETER              :: DeallocError = 'Deallocation error!'
   CHARACTER(Len=*), PARAMETER              :: AllocError   = 'Allocation error!'
   CHARACTER(Len=*), PARAMETER              :: LngOptErr    = 'Long option has no name!'
   CHARACTER(Len=*), PARAMETER              :: LevelError   = 'Wrong confidence level committed! Refer "samplestat --help"!'
   CHARACTER(Len=*), PARAMETER              :: OptError     = 'No meaningful options committed! Refer "samplestat --help"!'
   CHARACTER(Len=*), PARAMETER              :: R_OpnError   = 'Data file open error!'
   CHARACTER(Len=*), PARAMETER              :: R_ReadError  = 'Data file read error!'
   CHARACTER(Len=*), PARAMETER              :: R_AllocError = 'Data file allocation error!'

   !Report strings
   CHARACTER(Len=*), PARAMETER              :: Rpt_00       = '' ! left margin
   CHARACTER(Len=*), PARAMETER              :: Rpt_01       = 'sampleSTAT - Statistics for Sampling Distributions'
   CHARACTER(Len=*), PARAMETER              :: Rpt_02       = '=================================================='
   CHARACTER(Len=*), PARAMETER              :: Rpt_03       = 'Number of Values            : '
   CHARACTER(Len=*), PARAMETER              :: Rpt_04       = 'Arithmetic Mean             : '
   CHARACTER(Len=*), PARAMETER              :: Rpt_05       = 'Confidence Level            : '
   CHARACTER(Len=*), PARAMETER              :: Rpt_06       = 'Range of Dispersion (values): '
   CHARACTER(Len=*), PARAMETER              :: Rpt_07       = 'Range of Dispersion (mean)  : '
   CHARACTER(Len=*), PARAMETER              :: Rpt_08       = 'Standard Deviation          : '
   CHARACTER(Len=*), PARAMETER              :: Rpt_09       = 'Minimum                     : '
   CHARACTER(Len=*), PARAMETER              :: Rpt_10       = 'Maximum                     : '
   CHARACTER(Len=*), PARAMETER              :: Level0Msg    = '95%'
   CHARACTER(Len=*), PARAMETER              :: Level1Msg    = '99%'
   CHARACTER(Len=*), PARAMETER              :: Level2Msg    = '99.9%'

   ! Nullify pointer(s)
   NULLIFY(Values_ptr)

! --- Command-line parsing
   arg_nbr=command_argument_count() ! [nbr] Number of command line arguments
   arg_idx=1 ! [idx] Counting index

   cmd_ln: do while (arg_idx <= arg_nbr)

      call ftn_getarg_wrp(arg_idx,arg_val) ! [sbr] Call getarg, increment arg_idx

      dsh_key1=arg_val(1:1) ! First character of option (e.g. -)
      dsh_key2=arg_val(1:2) ! First two characters of option (e.g. --)
      opt_lng=ftn_opt_lng_get(arg_val) ! [nbr] Length of option

      if (opt_lng <= 0) stop "Long option has no name"

      ! Handle long options
      lng_cmd: IF (dsh_key2 == '--') THEN
         opt_sng=arg_val(3:2+opt_lng) ! Option string without --

         opt: IF (opt_sng == 'level' ) THEN         ! confidence level (/ 0, 1, 2 /)
            CALL ftn_arg_get(arg_idx,arg_val,Level)
            LogLevel = .TRUE.
         ELSE IF (opt_sng == 'version' ) THEN      ! Print version
            LogVer = .TRUE.
         ELSE IF (opt_sng == 'help' ) THEN         ! Print help page
            LogHlp = .TRUE.
         ELSE opt ! Option not recognized
            arg_idx=arg_idx-1 ! [idx] Counting index
            CALL ftn_getarg_err(arg_idx,arg_val) ! [sbr] Error handler for getarg()
         END IF opt ! endif option is recognized

         ! Jump to top of while loop
         CYCLE cmd_ln! C, F77, and F90 use 'continue', 'goto', and 'cycle'

      END IF lng_cmd ! endif long option

      ! Short option with parameters (e.g. -l 2)
      s_par: if (dsh_key2 == '-l') then
         call ftn_arg_get(arg_idx, arg_val, Level) ! Read parameter of '-l'
         LogLevel = .TRUE.
         cycle cmd_ln ! Next option
      ELSE IF (dsh_key2 == '/?') THEN
         LogHlp = .TRUE.
      end if s_par


      ! Handle short options
      short_cmd: IF ((dsh_key1 == "-") .AND. (dsh_key2 /= '--')) THEN
         DO i=2, opt_lng+1
            IF ((arg_val(i:i) == 'h') .OR. (arg_val(i:i) == 'H')) THEN
               LogHlp = .TRUE.
            ELSE IF ((arg_val(i:i) == 'v') .OR. (arg_val(i:i) == 'V')) THEN
               LogVer = .TRUE.
           ELSE
               arg_idx = arg_idx - 1
               call ftn_getarg_err(arg_idx, arg_val) ! Error handler
            end if
         end do
         cycle cmd_ln ! Next option
      END IF short_cmd

   END DO cmd_ln             ! end while (arg_idx <= arg_nbr)

! --- Examine committed options, read in and proceed data
   IF (arg_nbr == 0) THEN ! If no arguments are committed display version and help info
      CALL Help()
      STOP
   END IF

   ! Quit, if senseless options are committed ...


   ! Print help- or version-screen if right option was committed
   IF (LogHlp) CALL Help()
   IF (LogVer) CALL Version()
   IF (LogHlp .OR. LogVer) STOP ! Avoid further processing after print help or version

   ! Check committed options for "l" or "level"
   IF (( .NOT.(Level >= 0 .AND. Level <= 2)) .AND. ((Level .NE. 95) .AND. (Level .NE. 99) .AND. (Level .NE. 99.9))) THEN
       STOP LevelError
   END IF

   IF (LogLevel) THEN  !-------------------------------------------------------- Report
      ! Read in data
      CALL ReadInData(Values_ptr, N, ReadErr)
      ! Data file error handling
      IF (ReadErr == 1) STOP R_OpnError
      IF (ReadErr == 2) STOP R_ReadError
      IF (ReadErr == 3) STOP R_AllocError
      ! confidence level messages for report
      IF ((Level == 0) .OR. (Level == 95)) THEN ! Strayarea of the single values, stat. sec. 95%
         LevelMsg = Level0Msg
         StrayAreaResult = StrayArea(Values_ptr, N, 'lo')
      ELSE IF ((Level == 1) .OR. (Level == 99)) THEN ! Strayarea of the single values, stat. sec. 99%
         LevelMsg = Level1Msg
         StrayAreaResult = StrayArea(Values_ptr, N, 'md')
      ELSE IF ((Level == 2) .OR. (Level == 99.9)) THEN ! Strayarea of the single values, stat. sec. 99,9%
         LevelMsg = Level2Msg
         StrayAreaResult = StrayArea(Values_ptr, N, 'hi')
      END IF
      ! Report message
      WRITE(*,*)
      WRITE(*,*) Rpt_00, Rpt_01                             ! Title
      WRITE(*,*) Rpt_00, Rpt_02                             ! Underline
      WRITE(*,'(4A,i0)') ' ', Rpt_00, Rpt_03, '   ', N      ! Numbers of values
      WRITE(*,*) Rpt_00, Rpt_04, ArithMean(Values_ptr, N)   ! Mean
      WRITE(*,'(5A)') ' ', Rpt_00, Rpt_05, '   ', TRIM(LevelMsg)   ! confidence level
      WRITE(*,*) Rpt_00, Rpt_06, StrayAreaResult            ! Strayarea of the single values
      WRITE(*,*) Rpt_00, Rpt_07, StrayAreaResult/SQRT(REAL(N,KIND=DP))    ! TrustArea, Strayarea of mean
      WRITE(*,*) Rpt_00, Rpt_08, StdDev(Values_ptr, N)      ! Standard deviation
      WRITE(*,*) Rpt_00, Rpt_09, MinVar(Values_ptr, N)      ! Minimum
      WRITE(*,*) Rpt_00, Rpt_10, MaxVar(Values_ptr, N)      ! Maximum
      WRITE(*,*)                                            ! Blank line after report
   ELSE
      STOP OptError
   END IF

END PROGRAM sampleSTAT
